documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,486 @@
+- For dated DAGs, the existing reingestion DAGs will begin slowly backfilling
+  standardized popularity scores for historical records.
+
+[^1]:
+    There is one exception: iNaturalist does not use `get_record_data`, instead
+    doing its data processing in postgres. We do not need to do anything to
+    address this now because we do not track popularity data for iNaturalist;
+    _however_, if we want to add popularity data for it in the future we will
+    need to handle it separately.
+
+### Create a `popularity_refresh` DAG
+
+Once popularity calculation happens at ingestion (and assuming the media tables
+are backfilled -- more on that later), we need only handle the following cases:
+
+- A new popularity metric is added (for a new or existing provider)
+- The popularity constants are recalculated
+
+In both cases, we need to refresh the `<media>_popularity_constants` view (in
+order to recalculate constants). Note that once the constants have been updated,
+_the provider DAGs will immediately pick up the new constant and begin using it
+at ingestion_. All that remains is to update the scores for historical data.
+
+**Note**: It is true that our reingestion DAGs will also immediately begin
+updating the standardized scores of historical data using the new constant.
+However, this process is slow and some reingestion DAGs are currently running on
+sparse schedules. I think it is still worth it to explicitly updates the scores
+in this DAG, especially because that will allow us to use the DAG to do the
+initial backfill.
+
+As with our data refresh DAGs, we will use a factory to generate a
+`popularity_refresh` DAG for each media type. The DAG will do the following:
+
+1. Update the `<media>_popularity_metrics` table to include any newly added
+   metrics.
+1. Refresh the `<media>_popularity_constants` view to recalculate the popularity
+   constants.
+   1. This is done `CONCURRENTLY` so that provider DAGs can continue reading
+      from the view while it updates.
+1. For each unique `provider` in the `<media>_popularity_constants` view,
+   generate a `refresh_<provider>_scores` task. The task will run an `UPDATE` of
+   the `standardized_popularity` on all records matching that provider which
+   were last updated before the task began.
+   1. We may consider running the `refresh_<provider>_scores` tasks in parallel
+      to speed up the update.
+   1. Optionally, we can hard code a `SKIPLIST` of providers that are present in
+      the `<media>_popularity_constants` view, but for which we do not want to
+      create a refresh task. We currently have some providers (Nappy, Rawpixel,
+      Stocksnap) that support popularity data but are not dated, meaning scores
+      for **all** of their records will be updated the next time the DAG runs.
+      _Note that some of these DAGs are on a `@monthly` schedule however, which
+      means skipping them in this DAG could result in delayed recalculation
+      time._
```

Comment:
Exciting use case for [dynamic task mapping](https://airflow.apache.org/docs/apache-airflow/2.3.0/concepts/dynamic-task-mapping.html) :tada:

Reply:
Totally!